astrobase/periodbase/tls.py

```diff
@@ -0,0 +1,380 @@
+## IMPORTS ##
+#############
+
+import numpy as np
+
+from multiprocessing import Pool, cpu_count
+
+from math import fmod
+
+from numpy import (
+    nan as npnan, arange as nparange, array as nparray,
+    isfinite as npisfinite, argmax as npargmax, linspace as nplinspace,
+    ceil as npceil, argsort as npargsort, concatenate as npconcatenate
+)
+
+from astropy import units as u
+
+try:
+    from transitleastsquares import transitleastsquares
+except:
+    errmsg = (
+        'tried importing transitleastsquares and failed.\n'
+        'are you sure you have installed it correctly?\n'
+        'see https://transitleastsquares.readthedocs.io/en/latest/Installation.html'
+    )
+    raise AssertionError(errmsg)
+
+
+###################
+## LOCAL IMPORTS ##
+###################
+
+from ..lcmath import sigclip_magseries
+
+
+############
+## CONFIG ##
+############
+
+NCPUS = cpu_count()
+
+
+#######################
+## UTILITY FUNCTIONS ##
+#######################
+
+def tls_parallel_pfind(times, mags, errs=None,
+                       magsarefluxes=None,
+                       startp=0.1,  # search from 0.1 d to...
+                       endp=None,   # determine automatically from times
+                       tlsoversample=5,
+                       tlsmintransits=3,
+                       periodepsilon=0.1,
+                       nbestpeaks=5,
+                       sigclip=10.0,
+                       verbose=True,
+                       nworkers=None):
+    """
+    Wrapper to Hippke & Heller (2019)'s "transit least squares", which is BLS,
+    but with a slightly better template (and niceties in the implementation).
+
+    A few comments:
+
+    * The time series must be in units of days.
+
+    * The frequency sampling Hippke & Heller (2019) advocate for is cubic in
+    frequencies, instead of linear. Ofir (2014) found that the
+    linear-in-frequency sampling (which is correct for sinusoidal signal
+    detection) isn't optimal for a Keplerian box signal. He gave an equation
+    for "optimal" sampling. `tlsoversample` is the factor by which to
+    oversample over that. The grid can be imported independently via
+
+        `from transitleastsquares import period_grid`
+
+    The spacing equations are given here:
+    https://transitleastsquares.readthedocs.io/en/latest/Python%20interface.html#period-grid
+
+    * The boundaries of the period search are by default 0.1 day to 99% the
+    baseline of times.
+
+    Parameters
+    ----------
+
+    times,mags,errs : np.array
+        The magnitude/flux time-series to search for transits.
+
+    magsarefluxes : bool
+        As-implemented, must be true.
+
+    startp,endp : float
+        The minimum and maximum periods to consider for the transit search.
+
+    tlsoversample : int
+        Factor by which to oversample the frequency grid.
+
+    tlsmintransits : int
+        Sets the `min_n_transits` kwarg for the `BoxLeastSquares.autoperiod()`
+        function.
+
```

Comment:
Should add in the rest of the kwargs you supply to the `transitleastsquares.model` class in the body of the function here so they're configurable.